mesh/mesh.hpp

```diff
@@ -223,6 +224,7 @@ class Mesh
    Array<int> be_to_edge;  // for 2D
    Table *bel_to_edge;     // for 3D
    Array<int> be_to_face;
+   mutable Table *face_to_elem;
```

Comment:
This creates a bit of a quandary. This variable is used to cache the table created by `Mesh::GetFaceToElementTable`. However, the documentation for `Mesh::GetFaceToElementTable` clearly states that this table must be deleted by the caller implying that it is not a copy of internally cached data. Furthermore `Mesh::GetFaceToElementTable` recreates this table everything time it is called whereas `Mesh::FindFaceNeighbors` checks for the cached copy and uses that when present.

The new `face_to_elem` table is handled properly in this PR so that is not the issue. The question is should we treat `face_to_element` in the same manner as `face_edge` and `edge_vertex` i.e. cached internally and not deleted by the caller? This could break existing code unless we deprecate the existing `Mesh::GetFaceToElementTable` and create a new API to access the cached table. Another possibility is to keep the existing interface but alter `Mesh::GetFaceToElementTable` so that it copies the cached table rather than recreating it.

I don't know the best course of action but I think we should discuss this.

Reply:
I see the issue as the inconsistency of `GetFaceToElementTable` and other functions like `GetEdgeVertexTable` and `GetFaceEdgeTable`. Only `GetFaceToElementTable` gives the caller ownership. Of course, this inconsistency existed before this PR, and as you mentioned, this PR handles `face_to_elem` properly.

I agree with your suggestion, that backwards compatibility can be preserved by copying the internal table `face_to_elem` in `GetFaceToElementTable` and returning a pointer to the copy. However, I don't see an advantage over the current implementation in this PR, which sets the internal `face_to_elem` table only if `FindFaceNeighbors` is called. Since `FindFaceNeighbors` is a new function, it is only used by the new miniapp. Of course, it could be used more in the future. My choice of the current implementation is based on the rare usage of `FindFaceNeighbors`, i.e. it makes a minimal change, only in the case that FindFaceNeighbors is used.

In summary, my view is that the current implementation is fine, and there isn't a serious problem here to address. It would be nice to make things more consistent, but backwards compatibility makes that challenging.

Reply:
I agree that this isn't a serious problem. I guess I'm mainly concerned with maintainability. For the benefit of future developers perhaps we could add some comments before the `face_to_element` variable in the header file explaining how it is used (or not) by `FindFaceNeighbors` and `GetFaceToElementTable`.

Reply:
I added some documentation explaining this situation for the 3 tables.